include/seastar/core/relabel_config.hh

```diff
@@ -0,0 +1,80 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2022 ScyllaDB
+ */
+#include <regex>
+
+namespace seastar {
+namespace metrics {
+
+/*!
+ * \brief a wrapper class around regex with the original expr
+ *
+ * regex does not contain the original expression, this wrapper class
+ * acts both as a string and as a regex.
+ */
+class relabel_config_regex {
+    std::string _regex_str;
+    std::regex _regex;
+public:
+    relabel_config_regex() = default;
+    relabel_config_regex(const std::string& expr) : _regex_str(expr), _regex(std::regex(expr)) {}
+    relabel_config_regex(const char* expr) : _regex_str(expr), _regex(std::regex(expr)) {}
+    const std::string& str() const noexcept {
+        return _regex_str;
+    }
+    const std::regex& regex() const noexcept {
+        return _regex;
+    }
+
+    relabel_config_regex& operator=(const char* expr) {
+        std::string str(expr);
+        return operator=(str);
+    }
+
+    relabel_config_regex& operator=(const std::string& expr) {
+        _regex_str = expr;
+        _regex = std::regex(_regex_str);
+        return *this;
+    }
+};
+
+/*!
+ * \brief a relabel_config allows changing metrics labels dynamically
+ *
+ * The logic is similar to Prometheus configuration
+ * This is how Prometheus entry looks like:
+ *  - source_labels: [version]
+      regex: '([0-9]+\.[0-9]+)(\.?[0-9]*).*'
+      replacement: '$1$2'
+      target_label: svr
+ */
+struct relabel_config {
+    enum class relabel_action {skip_when_empty, report_when_empty, replace, keep, drop, drop_label};
+    std::vector<std::string> source_labels;
+    std::string target_label;
+    std::string replacement = "${1}";
+    relabel_config_regex expr = "(.*)";
+    relabel_action action = relabel_action::replace;
+    std::string separator = ";";
+};
+
+relabel_config::relabel_action relabel_config_action(const std::string& action);
```

Comment:
What does this do?

Reply:
translate a string to the enum, becase the enum is no longer in a class, I moved this to a function instead of a method.
I didn't wanted to wrap the entire enum just for that function.
One day C++ will have a string to enum cast...